### PR TITLE
Upgrade steps worker

### PR DESCRIPTION
**Pull request by SimonRichardson**
The following implements the upgrade steps worker. Upgrade steps are code changes normally to fix configuration files or perform API requests before the full jujud agent is unlocked. The basic premise is that once the upgrade database has become unlocked, the API server can come up and upgrade steps can then be performed.

The upgrade steps worker in the state implementation melded both the machine and controller worker into one. This made it extremely hard to follow. The new domain implementation splits the machine and controller worker into independent workers. The machine worker's only concern is to run the upgrade steps at each start-up if required. The controller upgrade steps ensure that all controllers are coordinated before performing the upgrade steps. Once all upgrade steps are completed, then and only then is the upgrade lock released.

It should be noted that the controller worker only uses the db to check if upgrade steps are performed. This is done to ensure that there is no local state. Everything is driven by the watchers from the db. The catacomb abstraction provides a nice way to do this, so we lean heavily on that worker pattern.

If the upgrade workers (database and steps) are killed during the upgrading process, then we try to mark the upgrade as failed. As there is a restriction on how many concurrent upgrades can happen at once. To prevent jujud being in a state that requires manual intervention, we set it too failed and then on the next go around it can attempt to try again.

Lastly, I had to rework the upgrades package to not take into account the state package. All model migrations will now be performed by the schema migration type, thus elevating it of that concern.

## Checklist

- [x] Code style: imports ordered, good names, simple structure, etc
- [x] Comments saying why design decisions were made
- [x] Go unit tests, with comments saying what you're testing

## QA steps

For changes that don't upgrade the version, then this flow works:

```sh
$ juju bootstrap lxd test --build-agent
$ juju add-model default
$ juju upgrade-controller --build-agent
```

To test that the upgrade database worker is called, you need to apply the
following diffs:

```diff
diff --git a/environs/sync/sync.go b/environs/sync/sync.go
index fa63d84426..f536f5e434 100644
--- a/environs/sync/sync.go
+++ b/environs/sync/sync.go
@@ -316,12 +316,12 @@ func buildAgentTarball(
        builtVersion.Build = 0
        clientVersion := jujuversion.Current
        clientVersion.Build = 0
-       if builtVersion.Number.Compare(clientVersion) != 0 {
-               return nil, errors.Errorf(
-                       "agent binary %v not compatible with bootstrap client %v",
-                       toolsVersion.Number, jujuversion.Current,
-               )
-       }
+       // if builtVersion.Number.Compare(clientVersion) != 0 {
+       //      return nil, errors.Errorf(
+       //              "agent binary %v not compatible with bootstrap client %v",
+       //              toolsVersion.Number, jujuversion.Current,
+       //      )
+       // }
        fileInfo, err := f.Stat()
        if err != nil {
                return nil, errors.Errorf("cannot stat newly made agent binary archive: %v", err)
```

Then update the patch version in `version/version.go`

```diff
diff --git a/version/version.go b/version/version.go
index 33b01db1af..3f599876ed 100644
--- a/version/version.go
+++ b/version/version.go
@@ -18,7 +18,7 @@ import (
 // The presence and format of this constant is very important.
 // The debian/rules build recipe uses this value for the version
 // number of the release package.
-const version = "4.0-beta1"
+const version = "4.0-beta2"

 // UserAgentVersion defines a user agent version used for communication for
 // outside resources.
```

```sh
$ juju bootstrap lxd test --build-agent
$ juju add-model default
$ juju upgrade-controller --build-agent
```

## Links

**Jira card:** JUJU-4757

### CODE REVIEW

File: core/watcher/eventsource/consume.go [hunk 1]
```diff
@@ -0,0 +1,81 @@
+// Copyright 2023 Canonical Ltd.
```
Comment (SimonRichardson) [inline core/watcher/eventsource/consume.go:1]: This file can be removed once the other PR lands and I can rebase ontop of it.

**Review by manadart — APPROVED** [review]
OK, this works. 

It has shortcomings that we've discussed, for which we will discuss solutions in Riga.

Let's iterate on it.

File: internal/database/bootstrap.go [hunk 2]
```diff
@@ -67,21 +67,51 @@ func (r *txnRunner) StdTxn(ctx context.Context, f func(context.Context, *sql.Tx)
...
 // provided BootstrapOpts on the specified model database.
 func BootstrapModelConcern(uuid model.UUID, ops ...BootstrapOpt) BootstrapConcern {
-	return bootstrapDBConcern(uuid.String(), schema.ModelDDL(), ops...)
+	return bootstrapDBConcern(uuid.String(), schema.ModelDDL(), EmptyInit, ops...)
+}
+
+// BootstrapControllerInitConcern is a BootstrapConcern type that will run the
+// provided BootstrapOpts on the controller database, after first running the
+// provided BootstrapInit.
+func BootstrapControllerInitConcern(bootstrapInit BootstrapInit, ops ...BootstrapOpt) BootstrapConcern {
```
[the remark this answers is not in the document]
Comment (SimonRichardson) [reply internal/database/bootstrap.go:89]: DummyCloud needs to be added before the options are added. There is a foreign key constraint that we violate somewhere if memory serves.

File: worker/machineupgradesteps/machineworker.go [hunk 3]
```diff
@@ -0,0 +1,131 @@
+// Copyright 2015 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package machineupgradesteps
```
[the remark this answers is not in the document]
Comment (SimonRichardson) [reply worker/machineupgradesteps/machineworker.go:4]: OK.

File: worker/upgradesteps/manifold.go [hunk 4]
```diff
@@ -1,34 +1,103 @@
...
-	"github.com/juju/juju/state"
+	"github.com/juju/juju/api/base"
+	"github.com/juju/juju/core/status"
+	"github.com/juju/juju/internal/servicefactory"
+	"github.com/juju/juju/internal/upgradesteps"
 	"github.com/juju/juju/upgrades"
 	"github.com/juju/juju/worker/gate"
 )
 
+type (
```
[the remark this answers is not in the document]
Comment (SimonRichardson) [reply worker/upgradesteps/manifold.go:22]: No. I'll remove them.

File: upgrades/preupgradesteps.go [hunk 5]
```diff
@@ -39,6 +35,21 @@ func PreUpgradeSteps(agentConf agent.Config, isController, isCaas bool) error {
 	return nil
 }
 
+// PreUpgradeStepsCAAS runs various checks for CAAS and prepares for
+// performing an upgrade. If any check fails, an error is returned which
+// aborts the upgrade.
+func PreUpgradeStepsCAAS(agentConf agent.Config, isController bool) error {
+	if isController {
```
[the remark this answers is not in the document]
Comment (SimonRichardson) [reply upgrades/preupgradesteps.go:42]: Correct.

File: cmd/jujud/main.go [hunk 6]
```diff
@@ -42,6 +42,7 @@ import (
 	"github.com/juju/juju/juju/osenv"
 	"github.com/juju/juju/juju/sockets"
 	_ "github.com/juju/juju/provider/all" // Import the providers.
+	"github.com/juju/juju/state"
```
[the remark this answers is not in the document]
Comment (SimonRichardson) [reply cmd/jujud/main.go:45]: Agreed.

I've no idea why it's not in core and wasn't something I was going to do in this PR.
